design/0001-frr.md

```diff
@@ -0,0 +1,347 @@
+# MLB-0001: BGP FRR enablement
+
+## Summary
+
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) to provide BGP functionality in MetalLB.
+When directed to, MetalLB will publish prefixes via FRR rather than
+MetalLB’s current built-in BGP speaker.
+
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+maintenance cost for the MetalLB project by reducing the effort required for
+adding additional BGP features and bug fixes. For example, there are a number of
+issues in the current backlog that may be addressed by using FRR:
+
+* [Add support for MP BGP encoding for IPv4 and IPv6](https://github.com/metallb/metallb/pull/590)
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+### Goals
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+
```

Comment:
FRR BGP configuration has more capabilities and options than what MetalLB BGP support today for example VRF support how do you plan to extend MetalLB to take advantage of those features ?

Reply:
@msherif1234 That could be a future enhancement but this enhancement just deals with achieving feature parity with the current implementation. All testing will reflect that.

Reply:
@msherif1234 I will add a non-goal to explicitly call this out.